File: Week02/590. N叉树的后序遍历.go
// 590. N叉树的后序遍历.go
package main

/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Children []*Node
 * }
 */

type Node struct {
	Val      int
	Children []*Node
}

//var aaa []int
//
//func postorder(root *Node) []int {
//	aaa = []int{}
//	dfs(root)
//	return aaa
//}
//
//func dfs(root *Node) {
//	if root != nil {
//		for _, n := range root.Children {
//			dfs(n)
//		}
//		aaa = append(aaa, root.Val) //后序输出
//	}
//}

func postorder(root *Node) []int {
	res := []int{}
	_dfs(root, &res)
	return res
}

func _dfs(root *Node, res *[]int) {
	if root != nil {
		for _, n := range root.Children {
			_dfs(n, res)
		}
		*res = append(*res, root.Val) //后序输出
	}
}

func postorder(root *Node) []int {
	var res []int
	if root == nil {
		return res
	}
	var stack = []*Node{root}
	for 0 < len(stack) {
		root = stack[len(stack)-1]
		res = append(res, root.Val)
		stack = stack[:len(stack)-1]
		l := len(root.Children)
		for i := 0; i < l; i++ {
			stack = append(stack, root.Children[i]) //入栈
		}
	}

	l := len(res) - 1
	for i := 0; i < l/2+1; i++ {
		res[i], res[l-i] = res[l-i], res[i]
	}
	return res
}

File: Week02/590. N叉树的后序遍历 .py
"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""

from typing import List


class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children


# 递归的写法
class Solution:
    def postorder(self, root: 'Node') -> List[int]:
        res = []
        self._helper(root, res)
        return res

    def _helper(self, root, res):
        if not root:
            return

        for node in root.children:
            self._helper(node, res)
        res.append(root.val)


# 迭代的写法

class Solution:
    def postorder(self, root: 'Node') -> List[int]:

        if not root:
            return []

        stack, output = [root, ], []
        while stack:
            root = stack.pop()
            if root is not None:
                output.append(root.val)
            for c in root.children:
                stack.append(c)

        return output[::-1]


# 以下为自我练习
class Solution:
    def postorder(self, root: 'Node') -> List[int]:
        if not root:
            return []
        # 左右根，根右左
        res = []
        stack = [root]
        while stack:
            root = stack.pop()
            res.append(root.val)
            stack.extend(root.children)

        return list(reversed(res))


class Solution:
    def postorder(self, root: 'Node') -> List[int]:
        if not root:
            return []

        stack = [root]
        res = []
        while stack:
            root = stack.pop()
            res.append(root.val)
            stack.extend(root.children)
        return res[::-1]


class Solution:
    def postorder(self, root: 'Node') -> List[int]:
        res = []
        self._helper(root, res)
        return res

    def _helper(self, root, res):
        if not root:
            return
        for child in root.children:
            self._helper(child, res)
        res.append(root.val)


class Solution:
    def postorder(self, root: 'Node') -> List[int]:
        if not root:
            return []
        res = []
        stack = [root]
        while stack:
            root = stack.pop()
            res.append(root.val)
            if root.children:
                stack.extend(root.children)
        return res[::-1]
